modules/core/src/nan_mask.simd.hpp

```diff
@@ -0,0 +1,439 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html
+
+
+#include "precomp.hpp"
+
+namespace cv {
+
+typedef void (*PatchNanFunc)(uchar* tptr, size_t len, double newVal);
+typedef void (*FiniteMaskFunc)(const uchar *src, uchar *dst, size_t total);
+
+CV_CPU_OPTIMIZATION_NAMESPACE_BEGIN
+
+PatchNanFunc getPatchNanFunc(bool isDouble);
+FiniteMaskFunc getFiniteMaskFunc(bool isDouble, int cn);
+
+#ifndef CV_CPU_OPTIMIZATION_DECLARATIONS_ONLY
+
+static void patchNaNs_32f(uchar* ptr, size_t ulen, double newVal)
+{
+    CV_INSTRUMENT_REGION();
+
+    int32_t* tptr = (int32_t*)ptr;
+    Cv32suf val;
+    val.f = (float)newVal;
+
+    int j = 0;
+    int len = (int)ulen;
+
+#if (CV_SIMD || CV_SIMD_SCALABLE)
+    v_int32 v_pos_mask = vx_setall_s32(0x7fffffff), v_exp_mask = vx_setall_s32(0x7f800000);
+    v_int32 v_val = vx_setall_s32(val.i);
+
+    int cWidth = VTraits<v_int32>::vlanes();
+    for (; j < len - cWidth*2 + 1; j += cWidth*2)
+    {
+        v_int32 v_src0 = vx_load(tptr + j);
+        v_int32 v_src1 = vx_load(tptr + j + cWidth);
+
+        v_int32 v_cmp_mask0 = v_lt(v_exp_mask, v_and(v_src0, v_pos_mask));
+        v_int32 v_cmp_mask1 = v_lt(v_exp_mask, v_and(v_src1, v_pos_mask));
+
+        if (v_check_any(v_or(v_cmp_mask0, v_cmp_mask1)))
+        {
+            v_int32 v_dst0 = v_select(v_cmp_mask0, v_val, v_src0);
+            v_int32 v_dst1 = v_select(v_cmp_mask1, v_val, v_src1);
+
+            v_store(tptr + j, v_dst0);
+            v_store(tptr + j + cWidth, v_dst1);
+        }
+    }
+#endif
+
+    for (; j < len; j++)
+    {
+        if ((tptr[j] & 0x7fffffff) > 0x7f800000)
+        {
+            tptr[j] = val.i;
+        }
+    }
+}
+
+
+static void patchNaNs_64f(uchar* ptr, size_t ulen, double newVal)
+{
+    CV_INSTRUMENT_REGION();
+
+    int64_t* tptr = (int64_t*)ptr;
+    Cv64suf val;
+    val.f = newVal;
+
+    int j = 0;
+    int len = (int)ulen;
+
+#if (CV_SIMD || CV_SIMD_SCALABLE)
+    v_int64 v_exp_mask = vx_setall_s64(0x7FF0000000000000);
+    v_int64 v_pos_mask = vx_setall_s64(0x7FFFFFFFFFFFFFFF);
+    v_int64 v_val = vx_setall_s64(val.i);
+
+    int cWidth = VTraits<v_int64>::vlanes();
+    for (; j < len - cWidth * 2 + 1; j += cWidth*2)
+    {
+        v_int64 v_src0 = vx_load(tptr + j);
+        v_int64 v_src1 = vx_load(tptr + j + cWidth);
+
+        v_int64 v_cmp_mask0 = v_lt(v_exp_mask, v_and(v_src0, v_pos_mask));
+        v_int64 v_cmp_mask1 = v_lt(v_exp_mask, v_and(v_src1, v_pos_mask));
+
+        if (v_check_any(v_cmp_mask0) || v_check_any(v_cmp_mask1))
+        {
+            // v_select is not available for v_int64, emulating it
+            v_int64 v_dst0 = v_or(v_and(v_cmp_mask0, v_val), v_and(v_not(v_cmp_mask0), v_src0));
+            v_int64 v_dst1 = v_or(v_and(v_cmp_mask1, v_val), v_and(v_not(v_cmp_mask1), v_src1));
```

Comment:
> // v_select is not available for v_int64, emulating it

reinterpret + vselect should work faster than provided emulation.

BTW, it makes sense to provide such implementation in a single place (HAL) /cc @vpisarev